### PR TITLE
extended meta API for batch tracking

### LINKED ISSUE #21: Extend frame and object meta API for batch tracking
Add properties to Savant meta API that are helpful for batch tracking in PyFunc

- [x] frame.source_id
- [x] frame.is_initial
- [x] object.uid

### PR DESCRIPTION
**Pull request by abramov-oleg**
Closes #21

### CODE REVIEW

File: savant/deepstream/pipeline.py [hunk 1]
```diff
@@ -298,8 +289,8 @@ def _on_source_caps(
                 time.sleep(5)
 
         with self._source_adding_lock:
-            self._source_id_by_index[source_info.pad_idx] = source_info.source_id
-            self._sources[source_info.source_id] = source_info
+            self._sources.set_src_id(source_info.pad_idx, source_info.source_id)
+            self._sources.set_src_info(source_info.source_id, source_info)
```
Comment (denisvmedyantsev) [inline savant/deepstream/pipeline.py:293]:
```
self._sources.set_src_id(source_info.pad_idx, source_info.source_id)
self._sources.set_src_info(source_info.source_id, source_info)
```
-> `self._sources.update_source(source_info)` or `self._sources.update_source(source_info.source_id, source_info)` 
there is no need to split this operation

File: savant/deepstream/pipeline.py [hunk 2]
```diff
@@ -468,7 +459,7 @@ def on_muxer_sink_pad_eos(self, pad: Gst.Pad, event: Gst.Event, source_id: str):
         self._logger.debug(
             'Got EOS on pad %s.%s', pad.get_parent().get_name(), pad.get_name()
         )
-        source_info = self._sources[source_id]
+        source_info = self._sources.get_src_info(source_id)
```
Comment (denisvmedyantsev) [inline savant/deepstream/pipeline.py:462]: just `self._sources.get_source(source_id)`?

File: savant/deepstream/pipeline.py [hunk 3]
```diff
@@ -558,8 +549,10 @@ def _remove_output_elems(self, source_info: SourceInfo):
         self._logger.debug(
             'Output elements for source %s removed', source_info.source_id
         )
-        del self._source_id_by_index[source_info.pad_idx]
-        del self._sources[source_info.source_id]
+
+        self._sources.free_src_id(source_info.source_id)
+        self._sources.free_pad_idx(source_info.pad_idx)
```
Comment (denisvmedyantsev) [inline savant/deepstream/pipeline.py:554]:
```
self._sources.free_src_id(source_info.source_id)
self._sources.free_pad_idx(source_info.pad_idx)
```
-> `self._sources.remove_source(source_info)` or `self._sources.remove_source(source_info.source_id)`?

File: savant/deepstream/pipeline.py [hunk 4]
```diff
@@ -620,7 +613,7 @@ def prepare_input(self, buffer: Gst.Buffer):
         self._logger.debug('Preparing input for buffer with PTS %s.', buffer.pts)
         nvds_batch_meta = pyds.gst_buffer_get_nvds_batch_meta(hash(buffer))
         for nvds_frame_meta in nvds_frame_meta_iterator(nvds_batch_meta):
-            source_id = self._source_id_by_index[nvds_frame_meta.pad_index]
+            source_id = self._sources.get_src_id(nvds_frame_meta.pad_index)
```
Comment (denisvmedyantsev) [inline savant/deepstream/pipeline.py:616]: -> `self._sources.get_source_by_pad_index(nvds_frame_meta.pad_index)` - a little more clear, IMHO

**Review by denisvmedyantsev — CHANGES_REQUESTED** [review]

File: savant/deepstream/meta/frame.py [hunk 5]
```diff
@@ -21,6 +22,18 @@ def __init__(
         self.batch_meta = frame_meta.base_meta.batch_meta
         self.frame_meta = frame_meta
 
+    @property
+    def source_id(self) -> str:
+        """Source id for the frame in the batch."""
+        return SourceInfoRegistry().get_src_id(self.frame_meta.pad_index)
+
+    @property
+    def is_initial(self) -> bool:
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply savant/deepstream/meta/frame.py]: Changed to frame_num property.

File: savant/deepstream/pipeline.py [hunk 6]
```diff
@@ -108,6 +97,9 @@ def __init__(
         # model-object association storage
         self._model_object_registry = ModelObjectRegistry()
 
+        self._source_adding_lock = Lock()
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply savant/deepstream/pipeline.py:100]: Left as is, seems like this lock is strongly related to pipeline modification operations.

File: savant/deepstream/pipeline.py [hunk 7]
```diff
@@ -244,22 +232,25 @@ def on_source_added(  # pylint: disable=unused-argument
...
             )
-        self._sources[source_id] = source_info
+        else:
+            while not source_info.lock.wait(5):
+                self._logger.debug(
+                    'Waiting source %s to release', source_info.source_id
+                )
+            source_info.lock.clear()
+
+        self._sources.set_src_info(source_id, source_info)
```
[the remark this answers is not in the document]
Comment (abramov-oleg) [reply savant/deepstream/pipeline.py]: Added the method.

File: savant/deepstream/pipeline.py (hunk 2 above)
Comment (abramov-oleg) [reply savant/deepstream/pipeline.py]: Renamed.

File: savant/deepstream/pipeline.py (hunk 3 above)
Comment (abramov-oleg) [reply savant/deepstream/pipeline.py]: Refactored.

File: savant/deepstream/pipeline.py (hunk 4 above)
Comment (abramov-oleg) [reply savant/deepstream/pipeline.py]: Renamed.

**Review by denisvmedyantsev — APPROVED** [review]

**Pull request merged**